bhyve.pyx

```diff
@@ -1,4 +1,139 @@
-cimport bhyve
+cimport vmmapi
+cimport cpuset
+cimport types
+cimport vmm
 
+import enum
+import os
 
+from libc.errno cimport errno
+from libc.stdlib cimport free
 
+
+class Error(enum.IntEnum):
+    OPEN_FAILED = 100
+    STATS_FETCH_FAILED = 101
+    RETRIEVE_ACTIVE_VCPU_FAILED = 102
+
+
+class BhyveException(RuntimeError):
+    def __init__(self, code, message):
+        super(BhyveException, self).__init__(f'[Errorno {code}] {message}')
+        self.code = code
+
+cdef class VM:
+
+    cdef str vm_name
+    cdef vmmapi.vmctx* vm
+
+    def __cinit__(self, str vm_name_p, bint create_vm=0):
+        self.vm_name = vm_name_p
+        vm_name_bytes = vm_name_p.encode()
+        cdef char* vm_name = vm_name_bytes
+
+        if create_vm:
+            with nogil:
+                code = vmmapi.vm_create(vm_name)
+
+            if code != 0:
+                raise BhyveException(
+                    errno, f'Failed to create VM {self.vm_name} - {os.strerror(errno)}'
+                )
+
+        with nogil:
+            self.vm = vmmapi.vm_open(vm_name)
+
+        if self.vm == NULL:
+            raise BhyveException(
+                Error.OPEN_FAILED, f'Could not open VM {self.vm_name}'
+            )
+
+    def force_reset(self):
+        with nogil:
+            error = vmmapi.vm_suspend(self.vm, vmm.VM_SUSPEND_RESET)
+
+        if error:
+            raise BhyveException(
+                errno, os.strerror(errno)
+            )
+
+        return True
+
+    def force_poweroff(self):
+        with nogil:
+            error = vmmapi.vm_suspend(self.vm, vmm.VM_SUSPEND_POWEROFF)
+
+        if error:
+            raise BhyveException(
+                errno, os.strerror(errno)
+            )
+
+        return True
+
+    def get_stats(self, vcpu=None):
+        active_vcpus = self.active_vcpus()
+        if vcpu is not None and not -1 < vcpu < active_vcpus:
+            raise BhyveException(
+                Error.STATS_FETCH_FAILED, f'Please specify a vcpu number from 0-{active_vcpus}'
```

Comment:
The error message here seems wrong, when we call this method we don't specify any vcpu, also vcpus should start from 1 to maximum 16 vcpus, although when you call libvmmapi, it will shows vcpu starting from 0 such like: 0, 1, 2, 3 for a guest with 4 vcpus.